### PR TITLE
SV1 Mining Device

### PR DESCRIPTION
**Pull request by rrybarczyk**
Mock SV1 Mining Device used by the translator proxy. Translator proxy PR #190 merges commits from this branch as needed. This is a separate PR into order to make things more manageable and separate out concerns.

### CODE REVIEW

File: examples/sv1-mining-device/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+/// Represents the Mining Device client which is connected to a Upstream node (either a SV1 Pool
+/// server or a SV1 <-> SV2 Translator Proxy server).
+pub(crate) struct Client {
+    client_id: u32,
+    extranonce1: HexBytes,
+    extranonce2_size: usize,
+    version_rolling_mask: Option<HexU32Be>,
+    version_rolling_min_bit: Option<HexU32Be>,
+    pub(crate) status: ClientStatus,
+    sented_authorize_request: Vec<(String, String)>, // (id, user_name)
```
Comment (ccdle12) [inline examples/sv1-mining-device/src/client.rs:32]: nit: typo? `sent_authorize_request`?

File: examples/sv1-mining-device/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+        // Initialize Client
+        let mut client = Client {
+            client_id,
+            extranonce1: "00000000".try_into().unwrap(),
+            extranonce2_size: 2,
+            version_rolling_mask: None,
+            version_rolling_min_bit: None,
+            status: ClientStatus::Init,
+            sented_authorize_request: vec![],
```
Comment (ccdle12) [inline examples/sv1-mining-device/src/client.rs:133]: nit: same typo `sent_authorize_request`

**Comment by xraid** [comment]
Test Coverage / Test Coverage (pull_request) Failing after 2m 

/stratum/protocols/v1/src/utils.rs line 10 `pub struct HexBytes(pub(crate) Vec<u8>);`
called from protocols/v1/src/methods/client_to_server.rs line 215
**barfs** because of (crate) makes struct private inside crate ? 
-- maybe add crate utils.rs in mod.rs or client_to_server.rs
-- or adjust utils.rs line 10 to `pub struct HexBytes(pub Vec<u8>);`

**Comment by xraid** [comment]
Test, Prop Tests, Example Tests / ci (macos-latest) (pull_request) Failing after 3m 

```
Run ./run.sh
  ./run.sh
  shell: /bin/bash -e {0}
+ cargo +nightly install cargo-fuzz
error: toolchain 'nightly-x86_64-apple-darwin' is not installed
Error: Process completed with exit code 1.
```

have all developers run stable to unify QA and simplify ...

**Comment by xraid** [comment]
Clippy Lint / clippy-check (ubuntu-latest) (pull_request) Failing after 2m 

Error: --> examples/sv1-mining-device/src/client.rs:72:5
pub(crate) async fn new(client_id: u32) {
= note: `-D clippy::new-ret-no-self` implied by `-D warnings`

```
pub trait Trait {
    // Good. Return type contains `Self`
    fn new() -> Self;
}
```

```
pub(crate) async fn new(client_id: u32) {
pub(crate) async fn new(client_id: u32) -> Self {
```
***

Error:    --> examples/sv1-mining-device/src/client.rs:240:13
```
match self.handle_message(message).unwrap() {
   Some(m) => {
        self.send_message(m).await;
   }
   None => (),
}
```
maybe could be changed to : 

```
if let Some(m) = self.handle_message(message).unwrap() {
     self.send_message(m).await;
   }
   None => (),
}
```

**Pull request merged**